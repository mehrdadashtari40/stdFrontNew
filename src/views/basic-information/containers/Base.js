import React, {useContext, useEffect} from 'react';

import 'react-dual-listbox/lib/react-dual-listbox.css';
import DynamicDataTable from '../components/DynamicDataTable';
import {connect, useSelector} from 'react-redux';
import {compose} from 'redux';
import * as base from '../_redux/baseInfoRedux';
import $ from 'jquery';
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import Spinner from "../../../common/Spinner/Spinner";
import {withTranslation} from "react-i18next";
import Select from 'react-select'
import {get_table_data, get_table_list} from "../_redux/baseInfoCrud";
import {AppConfig} from "../../../appConfig";

function Base(props) {
    const state = useSelector(state => state.basicInformation)
    const {apiServer} = useContext(AppConfig);
    useEffect(function () {
        if(state.tablesList.length === 0){
            get_table_list(apiServer).then(res=>{
                props.handle_variables({
                    tablesList:res
                })
            })
        }
    })

    const getTablesData = tableName => {
        props.handle_variables({loading:true});
        get_table_data(apiServer,tableName).then(res => {
            let primaryKey = '';
            res.structure.filter(item => {
                if (item.ai === 'yes') {
                    primaryKey = item.name;
                }
            });
            props.handle_variables({
                loading:false,
                data:res.data,
                structure:res.structure,
                selectedTable:tableName,
            })
        })
    };

    const onTableSelect = e => {
        const tableName = e.value;
        if (tableName) {
            getTablesData(tableName);
        } else {
            props.set_table_data({},'');
        }
    };

    const deleteRowItem = id => {
        var tableName = state.selectedTable;
        var name = getPrimaryKey();
        deleteItem(id, tableName, name);
    };

    const addNewItem = newItem => {
        props.setEmptyEditItem();
        props.addNewItem(newItem, () => {
            $('#editRowModal').modal('hide');
            getTablesData(state.selectedTable);
        });
    };

    const setEmptyValues = () => {
        if (state.data.length > 0) {
            var editRowData = state.data[0];
            for (var k in editRowData) {
                if (editRowData.hasOwnProperty(k)) {
                    const selector = '#' + k;
                    $(selector).val('');
                }
            }
        }
    };

    const onNewItem = () => {
        this.setEmptyValues();
        $('#editRowModal').modal('show');
    };

    const editRowItem = itemId => {
        props.editItem(itemId, state.primaryKey);
    };

    const updateItem = (id, data) => {
        var name = this.getPrimaryKey();
        props.updateItem(id, data, name, () => {
            $('#editRowModal').modal('hide');
        });
    };

    const getPrimaryKey = () => {
        let structure = state.structure;

        var name = '';
        structure.map(item => {
            if (item.ai == 'yes') {
                name = item.name;
            }
        });
        return name;
    };

    let dataTable = null;
    if (state.structure.length > 0) {
        let column = [];
        state.structure.map(item => {
            let obj = {
                class: '',
                data: null,
                render: function (data) {
                    return '<div>' + (data[item.name] ? data[item.name] : 'No data') + '</div>';
                },
            };
            column.push(obj);
        });
        dataTable = (
            <DynamicDataTable
                data={state.data}
                structure={state.structure}
                tableName={state.selectedTable}
                deleteRow={id => deleteRowItem(id)}
                column={column}
                addNew={newItem => addNewItem(newItem)}
                updateItem={(id, data) => updateItem(id, data)}
                editItem={id => editRowItem(id)}
                primaryKey={state.primaryKey}
            />
        );
    }


    if (dataTable === null && state.loading === false && state.tablesList.length !== 0) {
        getTablesData(state.tablesList[0].name)
    }
    let options = state.tablesList.map((t, index) => {
        return ({value: t.name, label: t.label});
    });
    return (
        <Card id="content" className={'overflow-visible'}>
            <CardContent>
                <h5 className={'table-header-title'}>
                    <i className={"fal fa-fw fa-database fa-2x"}></i>
                    مدیریت اطلاعات پایه
                    <div className={'base-table-select'}>
                        <Select
                            options={options}
                            onChange={onTableSelect} id="selected"/>
                    </div>
                    <button className="btn btn_inbox"
                            onClick={onNewItem}
                    >
                        <i className="fa fa-plus"/>
                        <span> درج اطلاعات</span>
                    </button>

                </h5>
                <div className="row">
                    <article className="col-sm-12 col-md-12 col-lg-12">
                        <div className="no-padding">
                            <div className="pull-right">

                            </div>
                            {state.loading === false ?
                                dataTable
                                :
                                <div
                                    style={{
                                        width: '100%',
                                        display: 'flex',
                                        justifyContent: 'center',
                                        alignItems: 'center',
                                    }}
                                >
                                    <Spinner/>
                                </div>
                            }
                        </div>
                    </article>
                </div>
            </CardContent>
        </Card>
    );
}

export default compose(withTranslation(),
    connect(null, base.actions))(Base);
