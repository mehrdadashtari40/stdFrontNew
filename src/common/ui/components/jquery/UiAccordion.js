import React, { useEffect, useRef } from "react";
import $ from "jquery";
import "smartadmin-plugins/es6/jquery-ui.min";

export default function UiAccordion({ children }) {
  const accordionRef = useRef(null);

  useEffect(() => {
    if (accordionRef.current) {
      $(accordionRef.current).accordion({
        autoHeight: false,
        heightStyle: "content",
        collapsible: true,
        animate: 300,
        icons: {
          header: "fa fa-plus",
          activeHeader: "fa fa-minus"
        },
        header: "h4"
      });
    }

    // Cleanup function to destroy accordion when component unmounts
    return () => {
      if (accordionRef.current) {
        $(accordionRef.current).accordion("destroy");
      }
    };
  }, []); // Empty dependency array = runs once on mount

  return (
    <div ref={accordionRef}>
      {children}
    </div>
  );
}