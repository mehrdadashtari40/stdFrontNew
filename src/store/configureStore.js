import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import { composeWithDevTools } from 'redux-devtools-extension';

import { userReducer, requestUserInfo } from '../common/user';
import { i18nReducer, i18nInit } from '../common/i18n';
import { trelloReducer } from '../views/trello';
import { systemManagementReducer } from '../views/system-management';
import { ProcessManagementReducer } from '../views/process-management';

import {reducer as layout2Reducer} from '../common/layout/_redux/layoutRedux'
import {reducer as tableReducer} from '../../src/common/tables/_redux/tableRedux'


import {reducer as basicInformationReducer} from '../views/basic-information/_redux/baseInfoRedux';
import LogsReducer from '../views/logs/LogReducesrs';

import BIReducer from '../views/BI/BIReducer';
import {reducer as inboxReducer} from "../views/inbox/_redux/inboxRedux";
import {reducer as authReducer} from '../../src/views/auth/_redux/authRedux'

import {reducer as ServiceDeskReducer} from "../views/service-desk/_redux/serviceDeskRedux";
import {reducer as SupervisorReducer} from "../views/supervisor/_redux/supervisorRedux";
import {reducer as CustomSettingsReducer} from "../views/customSettings/_redux/customSettingsRedux";
import {reducer as MenuReducer} from "../views/menu/_redux/menuRedux";
import {reducer as WorkDeskReducer} from "../views/work-desk/_redux/workDeskRedux";

export const rootReducer = combineReducers({
	auth: authReducer,
	user: userReducer,
	i18n: i18nReducer,
	basicInformation: basicInformationReducer,
	processManagement: ProcessManagementReducer,
	systemManagement: systemManagementReducer,
	ServiceDesk:ServiceDeskReducer,
	BI:BIReducer,
	trello:trelloReducer,
	inbox:inboxReducer,
	layout2:layout2Reducer,
	table:tableReducer,
	logs:LogsReducer,
	supervisor:SupervisorReducer,
	customSettingsReducer:CustomSettingsReducer,
	menuReducer:MenuReducer,
	workDeskReducer:WorkDeskReducer
});

const store = createStore(
	rootReducer,
	applyMiddleware(thunk)
);

store.dispatch(requestUserInfo());

store.dispatch(i18nInit());

export default store;
