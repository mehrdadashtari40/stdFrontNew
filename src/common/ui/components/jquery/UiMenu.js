import React, { useEffect, useRef } from "react";
import $ from "jquery";
import "smartadmin-plugins/es6/jquery-ui.min";

export default function UiMenu({ children }) {
  const menuRef = useRef(null);

  useEffect(() => {
    if (menuRef.current) {
      $(menuRef.current).menu();
    }

    // Cleanup function to destroy menu when component unmounts
    return () => {
      if (menuRef.current) {
        $(menuRef.current).menu("destroy");
      }
    };
  }, []); // Empty dependency array = runs once on mount

  return (
    <div ref={menuRef}>
      {children}
    </div>
  );
}