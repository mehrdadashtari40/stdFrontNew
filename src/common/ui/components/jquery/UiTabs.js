import React, { useEffect, useRef } from "react";
import $ from "jquery";
import "smartadmin-plugins/es6/jquery-ui.min";

export default function UiTabs(props) {
  const tabsRef = useRef(null);
  const { children, ...otherProps } = props;

  useEffect(() => {
    if (tabsRef.current) {
      $(tabsRef.current).tabs();
    }

    // Cleanup function to destroy tabs when component unmounts
    return () => {
      if (tabsRef.current) {
        $(tabsRef.current).tabs("destroy");
      }
    };
  }, []); // Empty dependency array = runs once on mount

  return (
    <div ref={tabsRef} {...otherProps}>
      {children}
    </div>
  );
}