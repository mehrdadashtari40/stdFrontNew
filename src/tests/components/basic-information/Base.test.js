import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import store from '../../../store/configureStore';
import configureStore from 'redux-mock-store';
import * as selectActions from '../../../views/new-inbox/NewInboxActions';
import Base from '../../../views/basic-information/containers/Base';
import $ from 'jquery'
jest.mock('react-i18next', () => ({
	withTranslation: () => Component => {
		Component.defaultProps = { ...Component.defaultProps, t: () => '' };
		return Component;
	},
}));

test('Make Snapshot file , Basic Information Component', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	expect(wrapper).toMatchSnapshot();
});


it('invokes `componentDidMount` when mounted', () => {
	jest.spyOn(Base.prototype, 'componentDidMount');
	shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	expect(Base.prototype.componentDidMount).toHaveBeenCalled();
	Base.prototype.componentDidMount.mockRestore();
});

test('getTablesData Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getTablesData).toBeDefined();
});
test('onTableSelect Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.onTableSelect).toBeDefined();
});
test('deleteRowItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.deleteRowItem).toBeDefined();
});
test('addNewItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.addNewItem).toBeDefined();
});
test('setEmptyValues Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.setEmptyValues).toBeDefined();
});


test('onNewItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.onNewItem).toBeDefined();
});


test('editRowItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.editRowItem).toBeDefined();
});

test('updateItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.updateItem).toBeDefined();
});


test('getPrimaryKey Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.getPrimaryKey).toBeDefined();
});


test('addNewItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.addNewItem).toBeDefined();
});

test('onNewItem Function  ', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();
	expect(instance.onNewItem).toBeDefined();
});



it('onTableSelect  work Corectly ?', () => {
	const wrapper = shallow(
		<Provider store={store}>
			<Base />
		</Provider>
	)
		.dive({ context: { store } })
		.dive();
	const instance = wrapper.instance();

	const spy = jest.spyOn(instance, 'onTableSelect');

	instance.forceUpdate();

	const select = wrapper.find('#selected');
	select.simulate('change', { target: { value: 'test' } });
	expect(spy).toHaveBeenCalledWith({ target: { value: 'test' } });
});