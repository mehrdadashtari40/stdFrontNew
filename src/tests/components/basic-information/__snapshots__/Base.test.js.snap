// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Make Snapshot file , Basic Information Component 1`] = `
<div
  id="content"
>
  <div
    className="row"
  >
    <BigBreadcrumbs
      className="col-xs-12 col-sm-7 col-md-7 col-lg-4"
      icon="fa fa-fw fa-circle-o-notch"
      items={
        Array [
          "مدیریت اطلاعات پایه ",
        ]
      }
    />
  </div>
  <WidgetGrid>
    <div
      className="row"
    >
      <article
        className="col-sm-12 col-md-12 col-lg-12"
      >
        <JarvisWidget
          collapsed={false}
          color={false}
          colorbutton={false}
          custombutton={false}
          deletebutton={false}
          editbutton={false}
          fullscreenbutton={true}
          hidden={false}
          load={false}
          refresh={false}
          sortable={true}
          togglebutton={false}
        >
          <div
            className="no-padding"
          >
            <h4>
              انتخاب جدول مورد نظر
            </h4>
            <br />
            <div
              className="pull-left"
            >
              <select
                id="selected"
                onChange={[Function]}
              />
            </div>
            <div
              className="pull-right"
            >
              <button
                className="btn btn-success"
                id="add-btn"
                onClick={[Function]}
                type="button"
              >
                درج اطلاعات
              </button>
            </div>
          </div>
        </JarvisWidget>
      </article>
    </div>
  </WidgetGrid>
</div>
`;
